File: C/FindAllAnagramsinaString.c
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

bool equalArray(int* shash, int* phash)
{
	for (int i = 0; i < 26; ++i)
	{
		if (shash[i] != phash[i])
			return false;
	}
	return true;
}

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* findAnagrams(char* s, char* p, int* returnSize)
{
	int shash[26] = {0}, phash[26] = {0};
	int slen = strlen(s), plen = strlen(p);
	*returnSize = 0;
	if (plen > slen)
		return NULL;
	for (int i = 0; i < plen; ++i)
		++phash[p[i] - 'a'];
	for (int i = 0; i < plen; ++i)
		++shash[s[i] - 'a'];
	int* res = (int*)malloc(sizeof(int) * 20101);
	if (equalArray(shash, phash))
		res[(*returnSize)++] = 0;
	for (int i = 0; i < slen - plen; ++i)
	{
		--shash[s[i] - 'a'];
		++shash[s[i + plen] - 'a'];
		if (equalArray(shash, phash))
			res[(*returnSize)++] = i + 1;
	}
	return res;
}

File: python/FindAllAnagramsinaString.py
from typing import List


class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        shash, phash, res = [0] * 26, [0] * 26, []
        slen, plen = len(s), len(p)
        if slen < plen:
            return res
        for c in p:
            phash[ord(c) - ord('a')] += 1
        for i in range(plen):
            shash[ord(s[i]) - ord('a')] += 1
        if shash == phash:
            res.append(0)
        for i in range(slen - plen):
            shash[ord(s[i]) - ord('a')] -= 1
            shash[ord(s[i+plen]) - ord('a')] += 1
            if shash == phash:
                res.append(i + 1)
        return res
